### PR TITLE
[WIP] [FEATURE] Add InferredAssetAzureDataConnector

### PR DESCRIPTION
**Pull request by fjork3**
Changes proposed in this pull request:
- https://superconductive.atlassian.net/browse/GREAT-76
- Implement `InferredAssetAzureDataConnector` class and related methods
-

### Definition of Done
Please delete options that are not relevant.

- [ ] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [ ] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [ ] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+from great_expectations.core.batch_spec import PathBatchSpec, AzureBatchSpec
+from great_expectations.exceptions.exceptions import ParserError
+
+from great_expectations.datasource.data_connector import (
+    InferredAssetFilePathDataConnector,
+)
+from great_expectations.datasource.data_connector.util import list_azure_keys
+from great_expectations.execution_engine import ExecutionEngine
+
+logger = logging.getLogger(__name__)
```
Comment (cdkini) [inline great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py]: Please add the `try/except` clause we have for cloud provider connection objects (see S3 or ConfiguredAssetAzure imports)

File: great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+
+class InferredAssetAzureDataConnector(InferredAssetFilePathDataConnector):
+    """
+    Extension of InferredAssetFilePathDataConnector used to connect to Azure blob store
```
Comment (cdkini) [inline great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py]: Please change all instances to Azure Blob Storage to align with official name

File: great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    The InferredAssetAzureDataConnector is one of two classes (ConfiguredAssetAzureDataConnector being the
+    other one) designed for connecting to filesystem-like data, more specifically files on Azure blob store. It
+    connects to assets inferred from container, prefix, and file name by default_regex.
+
```
Comment (cdkini) [inline great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py]: `prefix` should be `name_starts_with` to be in alignment with SDK

File: great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            sorters=sorters,
+            batch_spec_passthrough=batch_spec_passthrough,
+        )
+
+        self._container = container
+        self._name_starts_with = os.path.join(name_starts_with, "")
+        self._delimiter = delimiter
+
+        if azure_options is None:
+            azure_options = {}
```
Comment (cdkini) [inline great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py:84]: Please add comment and assertion from L94-101 of ConfiguredAssetAzureDataConnector

File: great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+        try:
+            if "conn_str" in azure_options:
+                self._azure = BlobServiceClient.from_connection_string(**azure_options)
+            else:
+                self._azure = BlobServiceClient(**azure_options)
+        except (TypeError, AttributeError):
+            raise ImportError(
+                "Unable to load Azure BlobServiceClient (it is required for InferredAssetAzureDataConnector)."
+            )
```
Comment (cdkini) [inline great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py]: Please follow the `try/except` logic from the ConfiguredAssetAzureDataConnector; we require regex to parse our `self._account_name` (which is used in `_get_full_file_path()`)

File: great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            recursive=True,
+        )
+        return path_list
+
+    def _get_full_file_path(
+        self,
+        path: str,
+        data_asset_name: Optional[str] = None,
+    ) -> str:
+        # data_assert_name isn't used in this method.
```
Comment (cdkini) [inline great_expectations/datasource/data_connector/inferred_asset_azure_data_connector.py]: Please fix typo to `asset`

**Review by cdkini — CHANGES_REQUESTED** [review]
Thanks for writing this up! Almost at the finish line 🎉 

I've requested some changes to the primary implementation; please take a look and let me know your thoughts.

Additionally, please add unit tests so we gain coverage over the changes made here. I very much modeled mine after the S3 ones (using `unitest.mock` to cover external dependencies like the anonymizer and Azure connection object). Happy to discuss live if that would be beneficial 😄

**Comment by alexsherstinsky** [comment]
Replaced by PR #3372

**Pull request closed without merging**